src/algebra/group/hom.lean

```diff
@@ -184,15 +242,208 @@ attribute [instance] inv.is_group_hom neg.is_add_group_hom
+attribute [to_additive add_monoid_hom.mk] monoid_hom.mk
+attribute [to_additive add_monoid_hom.mk.inj] monoid_hom.mk.inj
+attribute [to_additive add_monoid_hom.mk.inj_arrow] monoid_hom.mk.inj_arrow
+attribute [to_additive add_monoid_hom.mk.inj_eq] monoid_hom.mk.inj_eq
+attribute [to_additive add_monoid_hom.mk.sizeof_spec] monoid_hom.mk.sizeof_spec
+attribute [to_additive add_monoid_hom.no_confusion] monoid_hom.no_confusion
+attribute [to_additive add_monoid_hom.no_confusion_type] monoid_hom.no_confusion_type
+attribute [to_additive add_monoid_hom.rec] monoid_hom.rec
+attribute [to_additive add_monoid_hom.rec_on] monoid_hom.rec_on
+attribute [to_additive add_monoid_hom.sizeof] monoid_hom.sizeof
+attribute [to_additive add_monoid_hom.to_fun] monoid_hom.to_fun
+
+namespace monoid_hom
+variables {M : Type*} {N : Type*} {P : Type*} [monoid M] [monoid N] [monoid P]
+
+@[extensionality] def ext (f g : M →* N) (h : (f : M → N) = g) : f = g :=
+by cases f; cases g; cases h; refl
+
+/-- if f is a monoid homomorphism then f 1 = 1-/
+@[simp] lemma map_one (f : M →* N) : f 1 = 1 := f.map_one'
+
+/-- if f is a monoid homomorphism then f (a * b) = f a * f b -/
+@[simp] lemma map_mul (f : M →* N) (a b : M) : f (a * b) = f a * f b := f.map_mul' a b
+
+/-- the identity map from a monoid to itself-/
+def id (M : Type*) [monoid M] : M →* M :=
+{ to_fun := id,
+  map_one' := rfl,
+  map_mul' := λ _ _, rfl }
+
+/-- composition of monoid morphisms is a monoid morphism -/
+def comp (hnp : N →* P) (hmn : M →* N) : M →* P :=
+{ to_fun := hnp ∘ hmn,
+  map_one' := by simp,
+  map_mul' := by simp,
+}
+
+/-- The product of two monoids morphisms is a monoid morphism if the target is commutative-/
+def mul {M N} [monoid M] [comm_monoid N] (f g : M →* N) : M →* N :=
+{ to_fun := λ m, f m * g m,
+  map_one' := show f 1 * g 1 = 1, by simp,
+  map_mul' := begin intros, show f (x * y) * g (x * y) = f x * g x * (f y * g y),
+    rw [f.map_mul, g.map_mul, ←mul_assoc, ←mul_assoc, mul_right_comm (f x)], end
+}
+
+end monoid_hom
+
+namespace group_hom
+variables {G : Type*} [group G] {H : Type*} [group H]
+
+/-- makes a group homomomorphism from a proof that the map preserves multiplication -/
+def mk {f : G → H} (map_mul : ∀ a b : G, f (a * b) = f a * f b) : G →* H :=
+{ to_fun := f,
+  map_mul' := map_mul,
+  map_one' := mul_self_iff_eq_one.1 $ by rw [←map_mul, mul_one]
+}
+
+/-- makes a group homomorphism from a map and a proof that it preserves multiplication -/
+def mk' (f : G → H) (map_mul : ∀ a b : G, f (a * b) = f a * f b) : G →* H := mk map_mul
+
+/-- group homomorphisms preserve inverse -/
+theorem map_inv (f : G →* H) (g : G) : f g⁻¹ = (f g)⁻¹ :=
+eq_inv_of_mul_eq_one $ by rw [←f.map_mul, inv_mul_self, f.map_one]
+
+--@[to_additive add_group_hom.neg]
+/-- the inverse of a group homomorphism is a group homomorphism if the target is commutative-/
+def inv {G H} [group G] [comm_group H] (f : G →* H) : G →* H :=
```

Comment:
Idem dito, this should be in `monoid_hom` namespace, so that one can write `f.inv`. (Or maybe we put a `has_inv` instance on this set of morphisms. But still I think it wouldn't hurt to put it in the `monoid_hom` namespace.)

Reply:
OK!